### PR TITLE
Fix batch normalization in canned DNN Estimators

### PR DESCRIPTION
**Pull request by lukmanr**
This PR fixes the implementation of batch normalization in canned DNN estimators.  Batch normalization should be applied after the inputs but before the activation function.

### CODE REVIEW

**Comment by lukmanr** [comment]
This is a fix for issue #21915.

File: tensorflow/python/estimator/canned/dnn.py [hunk 1]
```diff
@@ -203,7 +214,21 @@ def _dnn_model_fn(features,
         batch_norm=batch_norm)
     logits = logit_fn(features=features, mode=mode)
 
-    if tpu_estimator_spec:
+    def batch_norm_train_op(loss):
```
Comment (ispirmustafa) [inline tensorflow/python/estimator/canned/dnn.py]: no need to have a different train_op

File: tensorflow/python/estimator/canned/dnn.py [hunk 2]
```diff
@@ -95,26 +97,35 @@ def dnn_logit_fn(features, mode):
...
               name='batchnorm_%d' % layer_id)
+          if activation_fn is not None:
+            net = activation_fn(net, name='hiddenlayer_%d_output' % layer_id)
+        else:
+          net = core_layers.dense(
+              net,
+              units=num_hidden_units,
+              activation=activation_fn,
+              kernel_initializer=init_ops.glorot_uniform_initializer(),
+              name=hidden_layer_scope)
```
Comment (ispirmustafa) [inline tensorflow/python/estimator/canned/dnn.py]: let's do following which will be more readable:
net = dense(activation=none)
if batch-norm:
  net = batch-norm(net)
if activation-fn:
  net = activation-fn(net)
if dropout:
  net = dropout

**Review by ispirmustafa — CHANGES_REQUESTED** [review]
Thank you Lukman.

**Review by ispirmustafa — CHANGES_REQUESTED** [review]
let's have an experimental configuration argument, without touching the default behavior. something like following:

_BatchNormOptions = namedtuple(...)
def experimental_batch_norm_options(
    apply_before_activation=False,
    momentum=0.999)
    return _BatchNormOptions(apply_before_activation=...)

DNNClassifier(..., batch_norm=experimental_batch_norm_options(...))

Adding wicke for comments

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by lukmanr** [comment]
Hi @ispirmustafa, the PR is updated per your suggestions.  I have tests for both options, apply_before_activation=False and apply_before_activation=True.

File: tensorflow/python/estimator/canned/dnn.py [hunk 3]
```diff
@@ -486,7 +519,7 @@ def __init__(
       config=None,
       warm_start_from=None,
       loss_reduction=losses.Reduction.SUM,
-      batch_norm=False,
+      batch_norm=None,
```
Comment (ispirmustafa) [inline tensorflow/python/estimator/canned/dnn.py]: This should also support boolean for backward compatibility. 
You can do something like following in init-fn:
if batch_norm and batchPnorm is bool:
  batchnorm = experimental_batch...

Comment (ispirmustafa) [reply tensorflow/python/estimator/canned/dnn.py]: Hi @lukmanr 
I couldn't see the backward compatibility. Could you please add a test to check backward compatibility: one with batch_norm=False, one with batch_norm=True.

**Comment by lukmanr** [comment]
Hi @ispirmustafa, backwards compatibility is enforced in the _dnn_logit_fn_builder method.  All the cases of batch_norm=False, batch_norm=True and batch_norm=_BatchNormOptions are now covered in the dnn tests, especially since I reverted to make batch_norm=False be the default in each of the canned estimator __init__ functions.

File: tensorflow/python/estimator/canned/dnn.py [hunk 4]
```diff
@@ -59,7 +81,9 @@ def _dnn_logit_fn_builder(units, hidden_units, feature_columns, activation_fn,
     dropout: When not `None`, the probability we will drop out a given
       coordinate.
     input_layer_partitioner: Partitioner for input layer.
-    batch_norm: Whether to use batch normalization after each hidden layer.
+    batch_norm: A boolean indicating whether to to use batch normalization
+      after each hidden layer with default options, or a `_BatchNormOptions`
```
Comment (ispirmustafa) [inline tensorflow/python/estimator/canned/dnn.py]: here instead of referring a private symbol '_BatchNormOptions', mention experimental_batch_norm_options

**Review by ispirmustafa — CHANGES_REQUESTED** [review]
Thank you!
one final minor comment. otherwise LGTM

**Comment by lukmanr** [comment]
Thanks - comment adjustments done.

**Review by ispirmustafa — APPROVED** [review]
Thank you @lukmanr!

File: tensorflow/python/estimator/canned/dnn.py [hunk 5]
```diff
@@ -40,6 +41,27 @@
 _LEARNING_RATE = 0.05
 
 
+_BatchNormOptions = collections.namedtuple(
```
Comment (josh11b) [inline tensorflow/python/estimator/canned/dnn.py]: Make this ExperimentalBatchNormOptions, export it, and get rid of the method below. You can set the defaults for a namedtuple using e.g. https://stackoverflow.com/questions/11351032/namedtuple-and-default-values-for-optional-keyword-arguments or inheritance which would also allow you to set the docstring.

**Review by josh11b — CHANGES_REQUESTED** [review]

File: tensorflow/python/estimator/canned/dnn.py (hunk 5 above)
Comment (josh11b) [reply tensorflow/python/estimator/canned/dnn.py]: (This was from TF API review.)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 20 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 35 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 52 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 67 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rthadur** [comment]
@lukmanr gentle ping to rebase your branch

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR due to lack of activity, If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
